fs-api-blockio/src/System/FS/BlockIO/Serial.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE LambdaCase #-}
+
+module System.FS.BlockIO.Serial (
+    serialHasBlockIO
+  ) where
+
+import           Control.Concurrent.Class.MonadMVar
+import           Control.Monad.Class.MonadThrow
+import           System.FS.API
+import qualified System.FS.BlockIO.API as API
+import           System.FS.BlockIO.API (IOOp (..), IOResult (..))
+
+-- | IO instantiation of 'HasBlockIO', using serialised I\/O.
+serialHasBlockIO ::
+     (MonadThrow m, MonadMVar m, Eq h)
+  => HasFS m h
+  -> HasBufFS m h
+  -> m (API.HasBlockIO m h)
+serialHasBlockIO hfs hbfs = do
+  ctx <- initIOCtx (SomeHasFS hfs)
+  pure $ API.HasBlockIO {
+      API.close = close ctx
+    , API.submitIO = submitIO hfs hbfs ctx
+    }
+
+data IOCtx m = IOCtx { ctxFS :: SomeHasFS m, openVar :: MVar m CtxState }
+
+data CtxState = Open Word | Closing Word | Closed
```

Comment:
Ok, I think I see what is going on here.

So we're trying to implement the behaviour specified in the API:

> Using 'submitIO' after 'close' should thrown an 'FsError' exception.

And we're doing that here by counting to track exactly when a IOCtx finally gets closed, so we can throw exceptions.

I think we can simplify. We're currently tracking this closing state, and then having `submitIO` fail if the ctx is closed by the time `submitIO` is finishing up. This is a bit odd. Firstly note that it involves `submitIO` and `close` being called concurrently. The API comments don't say what happens in this case, but in particular doesn't require that it cause a failure. Consider also what the blockio-uring one does: it make `close` prevent new submissions, and blocks until existing submissions have completed. So existing submissions are not made to fail.

We can simplify if we follow the latter behaviour: not requiring existing submissions fail, just failing new submissions after close.

In the serial case there's no actual cleanup that needs to happen since there are no extra resources. The only thing we want to do is catch incorrect uses of `close` vs `submitIO`. For that I think we _only_ need an `MVar Bool` to distinguish the closed / not-closed state. We would check that on `submitIO` and set it in `close`.

Reply:
> And we're doing that here by counting to track exactly when a IOCtx finally gets closed, so we can throw exceptions.

New `submitIO` will throw an error on a `Closing` or `Closed` context. Existing ones don't throw an error, but we do record when all of them are done by switching from `Closing` to `Closed`.

I do agree we only need a `Bool`, and that the `CtxState` is keeping track of more info then we need